### DISCUSSION TITLE
Is there any way to get the time that every frame taken

### DISCUSSION THREAD

**Discussion by BlockyDeer** [charmbracelet/bubbletea#870] [Q&A]
Is there any way to get the time that every frame taken? I have read the code but I cannot find anything about that. It's just like deltatime in Unity Engine or some other game engine. I want my animation played at the same speed.

**Comment by meowgorithm** [comment]
Hi!

To step animations at say, 60fps you can use the [`Every`](https://pkg.go.dev/github.com/charmbracelet/bubbletea#Every) command. You can get the time delta by comparing timestamps between messages.

```go
type StepAnimMsg time.Time

// Step the animation every 1/60th of a second.
func stepAnim() Cmd {
    return Every(time.Second/60, func(t time.Time) Msg {
        return StepAnimMsg(t)
    })
}

func (m model) Init() Cmd {
    // Start ticking.
    return stepAnim()
}

func (m model) Update(msg Msg) (Model, Cmd) {
    switch msg.(type) {
    case StepAnimMsg:
        now := time.Now()
        delta := m.lastStep.Since(now) // here's the time delta
        m.lastStep = now
        return m, stepAnim() // return your Every command again to loop.
    }
    return m, nil
}
```

The nice thing about `Every` is that all your animations will step at the same interval thanks to the nuances of the Bubble Tea runtime.

If you're interested in either springs or particles the [Harmonica](https://github.com/charmbracelet/harmonica) library works great in terminals and also supports a notion of a time delta.

There are also a couple of Haramonica/Bubble Tea examples that should help you wrap your head around animations in Bubble Tea in general:

* [Simple spring animation](https://github.com/charmbracelet/harmonica/blob/master/examples/spring/tui/main.go)
* [Cellbuffer-based example](https://github.com/charmbracelet/bubbletea/blob/master/examples/cellbuffer/main.go) (click around with the mouse)

Let us know if you still have questions.

* * *

Side note: the default max FPS is 60. You can change it with [`WithFPS`](https://github.com/charmbracelet/bubbletea/blob/master/options.go#L205-L212). So if you wanted a deltatime of ~17ms you could set an FPS of 60 (which is actually the default). Note that `WithFPS` is currently on `master` and not in a release yet so you'll need to `go get github.com/charmbracelet/bubbletea@master`. Also note that the max FPS is capped at 120.
